### PR TITLE
feat(component):  implement `select all` for multi-select component

### PR DESCRIPTION
**Pull request by bc-apostoliuk**
## What?
Implement `select all` for multi-select component;

## Why?

To be able to select/unselect all options by one click.

## Screenshots/Screen Recordings

https://github.com/bigcommerce/big-design/assets/84462142/a8df07b7-0bd7-447a-a369-0c9aa48fcb99

## Testing/Proof

Locally + UT.

### CODE REVIEW

**Review by MichalDobry — APPROVED** [review]

File: packages/big-design/src/components/MultiSelect/MultiSelect.tsx [hunk 1]
```diff
@@ -280,7 +286,13 @@ export const MultiSelect = typedMemo(
...
       inputValue,
       itemToString: (option) => (option ? option.content : ''),
-      items: filteredOptions,
+      items: [
+        ...(selectAll
+          ? // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+            ([{ content: localization?.selectAll, value: 'select-all' }] as Array<SelectOption<T>>)
+          : []),
+        ...filteredOptions,
+      ],
```
Comment (icatalina) [inline packages/big-design/src/components/MultiSelect/MultiSelect.tsx:295]: 🤔 I see why you need the `as` here, but I wonder if we should actually keep `value` as a generic and rather just type it as `string`.

I'm not sure it makes sense for value to be anything other than `string`, someone with more expertise in BD might correct me.

cc: @jorgemoya @chanceaclark

File: packages/big-design/src/components/List/List.tsx [hunk 2]
```diff
@@ -83,6 +90,61 @@ const StyleableList = typedMemo(
       scheduleUpdate();
     }, [isOpen, height, width, selectedItems?.length]);
 
+    const handleSelectAll = useCallback(
+      (listItems: Array<SelectOption<T>>) => {
+        const enabledItems = listItems.filter((item) => !item.disabled);
+
+        updateItems && updateItems(enabledItems);
```
[the remark this answers is not in the document]
Comment (bc-apostoliuk) [reply packages/big-design/src/components/List/List.tsx]: Make sense, Done 👍

File: packages/big-design/src/components/List/List.tsx [hunk 3]
```diff
@@ -83,6 +90,63 @@ const StyleableList = typedMemo(
...
+          const enabledItems = listItems.filter((item) => !item.disabled);
+
+          updateItems(enabledItems);
+        }
+      },
+      [updateItems],
+    );
+
+    const handleUnselectAll = useCallback(() => {
+      updateItems && updateItems([]);
```
Comment (icatalina) [inline packages/big-design/src/components/List/List.tsx:105]: the comment applied to all cases where we had an expression that wasn't really used. eg `a && a()` not just the one I pointed out

Comment (bc-apostoliuk) [reply packages/big-design/src/components/List/List.tsx]: My bad, missed that

Already fixed. Thanks.

File: packages/big-design/src/components/MultiSelect/MultiSelect.tsx (hunk 1 above)
Comment (jorgemoya) [reply packages/big-design/src/components/MultiSelect/MultiSelect.tsx:295]: I agree we should change this to string, I believe the original idea of having value as a generic is flawed, but maybe something we fix in a different PR.

File: packages/big-design/src/components/MultiSelect/spec.tsx [hunk 4]
```diff
@@ -1085,6 +1085,41 @@ test('renders localized labels', async () => {
   expect(label).toHaveStyleRule('content', "' (opcional)'", { modifier: '::after' });
 });
 
+test('selects all triggers onOptionsChange with available options', async () => {
```
Comment (jorgemoya) [inline packages/big-design/src/components/MultiSelect/spec.tsx:1088]: Can you add a test for indeterminate and unselect all?

Comment (bc-apostoliuk) [reply packages/big-design/src/components/MultiSelect/spec.tsx:1088]: Yes, added two more tests to cover these cases

**Review by jorgemoya — APPROVED** [review]

**Pull request merged**